server/src/d2Ast.ts

```diff
@@ -0,0 +1,202 @@
+        rngRet.push(new d2ExternalLink(n));
+      }
+    }
+
+    return rngRet;
+  }
+
+  /**
+   * Find all references, which is the sum of nodes and
+   * edges.  Since this may be used in multiple contexts,
+   * the values are cached so this loop doesn't have to run
+   * multiple times.
+   */
+  get References(): d2StringAndRange[] {
+    if (!this.references) {
+      this.references = [];
+
+      for (const node of this.nodes) {
+        // Edges
+        //
+        if (node.hasEdges) {
+          for (const edge of node.Edges) {
+            if (edge.src.edgeNode) {
+              this.references.push(edge.src.edgeNode);
+            }
+            if (edge.dst.edgeNode) {
+              this.references.push(edge.dst.edgeNode);
+            }
+          }
+        }
+
+        if (node.hasKey) {
+          if (node.Key?.key) {
+            this.references.push(node.Key.key);
+          }
+        }
+      }
+    }
+
+    return this.references;
+  }
+
+  /**
+   * Given a reference, return all references that are the same
+   */
+  public findAllMatchingReferences(ref: d2StringAndRange): d2StringAndRange[] {
+    const retArray: d2StringAndRange[] = [];
+    for (const r of this.References) {
+      if (r.str === ref.str) {
+        retArray.push(r);
+      }
+    }
+    return retArray;
+  }
+
+  /**
+   * Of all the references, which one is at the given position
+   */
+  public refAtPosition(pos: Position): d2StringAndRange | undefined {
+    for (const r of this.References) {
+      // TODO: Does this need to be more specific like nodeAtPosition?
+      const [isNode] = r.isPositionInRange(pos);
+      if (isNode) {
+        return r;
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * Get's the inner most node that is at the given position
+   */
+  public nodeAtPosition(pos: Position): d2Node | undefined {
+    let topScore = Number.MAX_SAFE_INTEGER;
+    let nRet;
+
+    for (const n of this.nodes) {
+      const [isNode, score] = n.isPositionInRange(pos);
+
+      if (isNode && score < topScore) {
+        topScore = score;
+        nRet = n;
+      }
+    }
+    return nRet;
+  }
+
+  /**
+   * Go through all the ast's node objects and break them
+   * apart to be used by the language server
+   */
+  private doNodes(nodes: LSPAny[]) {
+    for (const node of nodes ?? []) {
+      // console.log("---------------------------------");
+      // console.log(JSON.stringify(node, null, 2));
+      // console.log("---------------------------------");
+      const n = new d2Node(node);
+      this.nodes.push(n);
+
+      // NodeValues are nodes too!
+      if (n.hasNodeValues) {
+        for (const nv of n.nodeValueNodes) {
+          this.nodes.push(nv);
+        }
+      }
+    }
+  }
+
+  dump(): void {
+    console.log(`\n---------\nASTREADER: ${this.range.toString()}\n---------\n`);
+
+    for (const n of this.nodes) {
+      console.log(n.toString());
+    }
+  }
+}
+
+/**
+ ***********************
```

Comment:
what's the point of these

Reply:
I had some test data after this and was trying to make it obvious, but its all gone now.